content/blog/2022-04-28-chlorophyll/index.md

```diff
@@ -0,0 +1,309 @@
+<center>
+<img src="perf.png" alt="alt_text" title="image_tooltip" style="zoom:50%;" />
+</center>
+
+They do compare their generated code with handwritten code and give the following statement:
+> "Compared to the experts' implementation, it is **only 65% slower, 70% less energy-efficient and uses 2.2x more cores**. This result confirms that our generated programs are comparable with experts’ not only on small programs but also on a real application."
+
+The authors are somehow proud of this result. Considering Chlorophyll is the first high-level compiler for GA144, if it indeed reduces the programming effort, then even the performance is not good, it is useful for more applications running on GA144.
+
+### Productivity
+The authors also mention that a graduate student spent one summer learning arrayForth to program GA144 but only implemented 2 benchmarks. With Chlorophyll, authors can implement 5 benchmarks within one afternoon, and the 2-core version is better than the expert's implementation.
+
+We all agree this coding sample is too small. The authors can hold an undergraduate class to gather more coding examples and compare the efficiency of their proposed compiler.
+
+### Compilation Time
+The authors also provide the compilation time of different benchmarks. We can see from the following figure, it takes more than 16 hours to compile the largest benchmark. Most of the time is spent in the superoptimization part.
+<center>
+<img src="compilation-time.png" alt="alt_text" title="image_tooltip" style="zoom:45%;" />
+</center>
+
+It seems the compilation time is not count into the development time mentioned above. If compiling a program costs such a long time, it is hard for a programmer to debug. It is also impossible to write a program and verify its correctness in just one afternoon.
+
+
+## Further Discussion
+### Synthesis and Compilation
+Prof. Armando Solar-Lezama at MIT mentions in his course, "One distinguishing feature between a compiler and a synthesizer is the element of **search**"[^3]. Traditional compiler only does some simple transformation to the program, while synthesizer may have lots of valid programs but need to search for the best one. However, we can see the boundary between compiler and synthesizer nowadays is becoming blurry. A trend of domain-specific compiler is to decouple the *schedule* (how to optimize the program) from the *algorithm* (what to compute). Using this new interface, some compilers may be equipped with auto-tuning methods, which is essentially what the synthesizer does for searching. Examples include Halide[^4] for image processing, TVM[^5] for deep learning, HeteroCL[^6] for software-defined hardware, and GraphIt[^7] for graph processing, etc.
```

Comment:
Good point; referencing these "compilers optimized for search" projects is useful commentary.